docs/styleguides/python.md

```diff
@@ -0,0 +1,173 @@
+
+        ```
+        Use import x for importing packages and modules.
+        Use from x import y where x is the package prefix and y is the module name with no prefix.
+        Use from x import y as z if two modules named y are to be imported or if y is an inconveniently long name.
+        ```
+    * Use `__init__.py` only to represent python packages.
+        * Avoid importing of objects in `__init__.py` for readability.
+* Import formatting
+
+    * The types of imports should be separated by single line and order is should be.
+        * standard library imports.
+        * third-party imports.
+        * application specific imports.
+
+    * The `from` style import and `import` statements should be separated by single line too.
+    * Each group of these imports should be sorted.
+
+For Example
+
+```
+import json
+import uuid
+
+from subprocess import CalledProcessError
+from typing import Callable, List
+
+import retrying
+
+from retrying import retry, RetryError
+from teamcity.messages import TeamcityServiceMessages
+
+import test_util.aws
+import test_util.cluster
+
+from pkgpanda.util import logger, write_string
+```
+
+* [Packages](https://google.github.io/styleguide/pyguide.html?showone=Packages#Packages)
+    * Import each module using the full pathname location of the module.
+    * No Relative Imports.
+
+* Maximum line length is 120 characters.
+    * Following are the exceptions to this rule. These types can be of arbitrary length.
+        * Long import statements.
+        * URLs in comments.
+        * Any sequence of chars which would better be in a single line.
+
+* Indentation
+
+    * Use 4space indentation.
+    * For long list of arguments, use double-indented hanging-indentation.
+ For e.g.
+    * The hanging indentation should be followed by a single blank line.
+
+ ```
+ class DcosApiSession(ARNodeApiClientMixin, ApiClientSession):
+
+     def __init__(
+             self,
+             dcos_url: str,
+             masters: list,
+             public_masters: list,
+             slaves: list,
+             public_slaves: list,
+             default_os_user: str,
+             auth_user: Optional[DcosUser]):
+
+         # ...
+         super().__init__(Url.from_string(dcos_url))
+         self.masters = sorted(masters)
+ ```
+
+
+* Adding Types. Since our code-base is Python3.5+ compatible, we have the choice of introducing types, validating with mypy.
+
+        class VpcClusterUpgradeTestDcosApiSessionFactory:
+
+            def apply(
+                    self,
+                    dcos_url: str,
+                    masters: List[str],
+                    public_masters: List[str],
+                    slaves: List[str],
+                    public_slaves: List[str],
+                    default_os_user: str) -> DcosApiSession:
+
+    * Reviewers will request the author to write types for the method parameters.
+    * We will start running mypy type checker as part of build process.
+
+* Shebang Rule.
```

Comment:
It would be useful to explain why this is. I believe it is the only consistent way across platforms to guarantee which Python version is required. In general the "why" behind changes is useful because later we can know whether that reason is still valid / outdated / it was always dubious, and we can change the guide accordingly.